File: archive/archive/asmlToAsm/ASML/Tree.ts
import { asmlCall } from './asmlCall';
import { asmlNew } from './asmlNew';
import { asmlMem } from './asmlMem';
import { asmlArithmetics } from './asmlArithmetics';
import { asmlLabel } from './asmlLabel';
import { asmlBranch } from './asmlBranch';
import { asmlFunc } from './asmlFunc';
import { asmlFloat } from './asmlFloat';
import { asmlIf } from './asmlIf';
import { asmlLet } from './asmlLet';
import { asmlFunDef } from './common/asmlFunDef';
import { insType } from './common/insType';

class Tree {
    labels: any[] = [];

    constructor(code: string) {
        const instructions = code.split('\n');
        let currentBranch = new asmlBranch();

        for (const instr of instructions) {
            currentBranch = this.decoderInstruction(instr.trim(), currentBranch);
        }
    }

    private decoderInstruction(instruction: string, currentBranch: any): any {
        const instrType = insType.getTypeInstruction(instruction);

        switch (instrType) {
            case insType.LET_FUN:
                const a_fun = new asmlFunc(instruction);
                this.labels.push(a_fun);
                currentBranch = a_fun;
                break;
            case insType.LET_FLOAT:
                const a_float = new asmlFloat(instruction);
                this.labels.push(a_float);
                break;
            case insType.IF:
                const a_if = new asmlIf(instruction);
                currentBranch.addInstruction(a_if);
                const tmp = currentBranch;
                currentBranch = a_if;
                currentBranch.setParent(tmp);
                break;
            case insType.ELSE:
                currentBranch.setConstructionThen(false);
                break;
            case insType.FI:
                currentBranch = currentBranch.getParent();
                break;
            case insType.LET_IN:
                const a_let = new asmlLet(instruction);
                currentBranch.addInstruction(a_let);
                break;
            case insType.ADD:
            case insType.SUB:
            case insType.FADD:
            case insType.FSUB:
                const a_arith = new asmlArithmetics(instruction);
                currentBranch.addInstruction(a_arith);
                break;
            case insType.CALL:
                const a_call = new asmlCall(instruction);
                currentBranch.addInstruction(a_call);
                break;
            case insType.NOP:
                // Do nothing
                break;
            case insType.MEM:
                const a_mem = new asmlMem(instruction);
                currentBranch.addInstruction(a_mem);
                break;
        }
        return currentBranch;
    }

    registerAllocation_Spill(): void {
        for (const label of this.labels) {
            if (label instanceof asmlFunc) {
                label.allocRegSpill();
            }
        }
    }

    generateAsm(): string {
        let code = "\t.text\n";
        code += "\t.global _start\n";
        for (const a of this.labels) {
            code += a.generateAsm();
        }
        return code;
    }

    toString(): string {
        let res = "-------- TREE ----------\n";
        for (const a of this.labels) {
            res += a.toString();
        }
        return res;
    }
}

export { Tree };

File: archive/archive/ASML2ASM/ASML/Tree.py
from .asmlCall import asmlCall
from .asmlMem import asmlMem
from .asmlArithmetics import asmlArithmetics
from .asmlBranch import asmlBranch
from .asmlFunc import asmlFunc
from .asmlFloat import asmlFloat
from .asmlIf import asmlIf
from .asmlLet import asmlLet
from .common.insType import insType


class Tree:
    def __init__(self,code):
        self.labels=[]
        instructions=code.splitlines()
        currentBranch=asmlBranch() 
        
        for instr in instructions:
            currentBranch=self.decoderInstruction(instr,currentBranch)

    def decoderInstruction(self, instruction,currentBranch):
        instruction=instruction.strip()
        instrType=insType.getTypeInstruction(instruction)

        if instrType==insType.LET_FUN:
            a_fun=asmlFunc(instruction)
            self.labels.append(a_fun)
            currentBranch=a_fun
        elif instrType==insType.LET_FLOAT:
            a_float=asmlFloat(instruction)
            self.labels.append(a_float)
        elif instrType==insType.IF:
            a_if=asmlIf(instruction)
            currentBranch.addInstruction(a_if)
            tmp=currentBranch
            currentBranch=a_if
            currentBranch.setParent(tmp)
        elif instrType==insType.ELSE:
            currentBranch.setConstructionThen(False)
        elif instrType==insType.FI:
            currentBranch=currentBranch.getParent()
        elif instrType==insType.LET_IN:
            a_let=asmlLet(instruction)
            currentBranch.addInstruction(a_let)
        elif instrType==insType.ADD:
            a_arith=asmlArithmetics(instruction)
            currentBranch.addInstruction(a_arith)
        elif instrType==insType.SUB:
            a_arith=asmlArithmetics(instruction)
            currentBranch.addInstruction(a_arith)
        elif instrType==insType.FADD:
            a_arith=asmlArithmetics(instruction)
            currentBranch.addInstruction(a_arith)
        elif instrType==insType.FSUB:
            a_arith=asmlArithmetics(instruction)
            currentBranch.addInstruction(a_arith)
        elif instrType==insType.CALL:
            a_call=asmlCall(instruction)
            currentBranch.addInstruction(a_call)
        elif instrType==insType.NOP:
            pass
        elif instrType==insType.MEM:
            a_mem=asmlMem(instruction)
            currentBranch.addInstruction(a_mem)
        return currentBranch

    def registerAllocation_Spill(self):
        for label in self.labels:
            if isinstance(label,asmlFunc):
                #parameters
                label.allocRegSpill()
    
    def generateAsm(self):
        code = ""
        code += "\t.text\n"
        code += "\t.global _start\n"
        for a in self.labels:
            code += a.generateAsm()
        return code
    
    def __str__(self):
        res = "-------- TREE ----------\n"
        for a in self.labels:
            res += str(a)
        return res
